### PR TITLE
Create deployment file directory so it's owned by the user

### LINKED ISSUE #547: the role 'kubernetes-modules' was not found
##### ISSUE TYPE

 - Documentation Report

##### SUMMARY
Posting this as an issue/question since I did not see any updates in the docs for Openshift deploy with 0.9 and am wondering how to proceed. After generating the `openshift.yml` playbook to deploy my services I immediately hit the following:

```
ansible-playbook ansible-deployment/openshift.yml
```

```
 [WARNING]: provided hosts list is empty, only localhost is available

ERROR! the role 'kubernetes-modules' was not found in /home/ehelms/workspace/openshift/forklift/openshift/ansible-deployment/roles:/home/ehelms/workspace/openshift/forklift/openshift/roles:/home/ehelms/workspace/openshift/playbooks/roles:/home/ehelms/workspace/openshift/forklift/openshift/ansible-deployment

The error appears to have been in '/home/ehelms/workspace/openshift/forklift/openshift/ansible-deployment/openshift.yml': line 7, column 9, but may
be elsewhere in the file depending on the exact syntax problem.

The offending line appears to be:

    roles:
      - role: kubernetes-modules
        ^ here
```

In 0.3, everything was included in the generated role so I wasn't sure if that was simply missing in 0.9 at the moment or if I am supposed to include this role manually via `requirements.yml`.

### LINKED ISSUE #557: ansible-deployment/ directory is owned by root:root
##### ISSUE TYPE

 - Bug Report

##### SUMMARY
When running `ansible-container --engine openshift deploy` an ansible-deployment/ directory is created along with a playbook for the openshift deployment (e.g. openshift.yml). Both of these are owned by root:root the first time they are created (I assume because they are created by the conductor container). When running as my normal user this is annoying because I cannot easily edit or install additional roles into that directory without having to update the perms.

### PR DESCRIPTION
**Pull request by ryansb**
Closes #557
Closes #547

##### ISSUE TYPE
 - Bugfix Pull Request

##### SUMMARY
- Previously the mounting of a nonexistent directory would create it, owned by root. Now we check first, and create it with the ownership of the invoking user. 
- Recursively checks files/folders within the `ansible-deployment` path, and sets the uid/gid to that of the user.
- Adds a `pre_deployment_setup` method to the engines, giving the engine an opportunity to do things like install Ansible roles required by the generated playbook. In the case of k8s/openshift, then ansible.kubernetes-modules role is installed.
- Updates `deploy` command reference doc

### CODE REVIEW

File: container/docker/engine.py [hunk 1]
```diff
@@ -231,6 +231,8 @@ def run_conductor(self, command, config, base_path, params, engine_name=None, vo
 
         if params.get('deployment_output_path'):
             deployment_path = params['deployment_output_path']
+            if not os.path.isdir(deployment_path):
+                os.path.mkdir(deployment_path, 0o755)
```
Comment (ehelms) [inline container/docker/engine.py]: s/os.path.mkdir/os.mkdir

**Comment by ehelms** [comment]
This fixes the permissions on the directory which helps with role installation into the directory, however, the playbook is still owned by root (see below). This does not prevent me from running the playbook but its a bit annoying if I want to open the file up or make any changes manually for testing.

```
ehelms@war openshift (openshift-2.0)$ ls -la ansible-deployment/
total 64
drwxr-xr-x 3 ehelms ehelms  4096 May 26 18:16 .
drwxrwxr-x 8 ehelms ehelms  4096 May 26 18:15 ..
drwxrwxr-x 8 ehelms ehelms  4096 May 26 18:16 ansible.kubernetes-modules
-rw-r--r-- 1 root   root   51157 May 26 18:16 foreman.yml
```

**Comment by ehelms** [comment]
I tested this in my local environment and it allowed my deploy workflow to pass without any modifications. So a big :+1: functionality wise.

File: docs/rst/reference/deploy.rst [hunk 2]
```diff
@@ -3,45 +3,81 @@ deploy
...
+The ``deploy`` command maps your ``container.yml`` file to a cloud configuration, depending on the engine. See the :doc:`../../container_yml/reference`
+for details on how directives are mapped, and for cloud specific options.
+
+The playbook name will match the name of the project, and have a ``.yml`` extension. The playbook is created in in the ``ansible-deployment`` directory, which will be 
+created automatically. Use the *deployment_output_path* option in the *settings* section of ``container.yml`` to write the playbook to a different directory.
+
+.. note::
+
+    For K8s and OpenShift, the generated playbook requires the ``ansible.kubernetes-modules`` role, which is automatically installed to ``ansible-deployment/roles``.
+    Running the playbook from the ``ansible-deployment`` directory allows the playbook to access the role.
```
Comment (ehelms) [inline docs/rst/reference/deploy.rst]: You can actually run it from your main directory, e.g.:

```
ansible-playbook ansible-deployment/openshift.yml --tags start
```

**Comment by ehelms** [comment]
I may have spoke too soon, I had not tested `run`:

```
Parsing conductor CLI args.
Engine integration loaded. Preparing run.	engine=Docker™ daemon
Verifying service image	service=pulp-base
Verifying service image	service=content-server
Traceback (most recent call last):
  File "/usr/bin/conductor", line 11, in <module>
    load_entry_point('ansible-container', 'console_scripts', 'conductor')()
  File "/_ansible/container/__init__.py", line 19, in __wrapped__
    return fn(*args, **kwargs)
  File "/_ansible/container/cli.py", line 356, in conductor_commandline
    **params)
  File "/_ansible/container/__init__.py", line 19, in __wrapped__
    return fn(*args, **kwargs)
  File "/_ansible/container/core.py", line 762, in conductorcmd_run
    rc = run_playbook(playbook, engine, services, tags=['start'], **kwargs)
  File "/_ansible/container/__init__.py", line 19, in __wrapped__
    return fn(*args, **kwargs)
  File "/_ansible/container/core.py", line 522, in run_playbook
    engine.pre_playbook_tasks(deployment_output_path=deployment_output_path, **kwargs)
AttributeError: 'Engine' object has no attribute 'pre_playbook_tasks'
Conductor terminated. Cleaning up.	command_rc=1 conductor_id=a7fe627e33da751057d8bdc5afdd84668c95a7f20923c55e6432ffa180ee8977 save_container=False
ERROR	Conductor exited with status 1	
```

**Comment by chouseknecht** [comment]
Removed that line before merging. It broke in tests as well, which prompted me to fix it.

**Comment by ehelms** [comment]
Ah great, I'll update.

**Pull request merged**